Airflow final fixes

valayDave opened the pull request.

valayDave committed:
commit 0
- unfinished code.

valayDave committed:
more tweeks.

valayDave committed:
First version of dynamic dag compiler.
- Not completely finished code
- Creates generic .py file a JSON that is parsed to create Airflow DAG.
- Currently only boiler plate to make a linear dag but doesn't execute anything.
-  Unfinished code.

valayDave committed:
some changes based on savin's comments.
- Added changes to task datastore for different reason : (todo) Decouple these
- Added comments to SFN for reference.
- Airflow DAG is no longer dependent on metaflow

valayDave committed:
Tweeks

valayDave committed:
Got linear workflows working on airflow.
- Still not feature complete as lots of args are still unfilled / lots of unknows
- minor tweek in eks to ensure airflow is k8s compatible.
- passing state around via xcom-push
- HACK : AWS keys are passed in a shady way. : Reverse this soon.

valayDave committed:
Fixed setting run-id / task-id to labels in k8s
- Fixed setting run-id has from cli macro
- added hashing macro to ensure that jinja template set the correct run-id to k8s labels
-

valayDave committed:
Added @schedule support + readme

valayDave committed:
Nit fix.

valayDave committed:
setting retry logic.

valayDave committed:
Fixed setting `task_id` :
- switch task-id from airflow job is to hash to "runid/stepname"
- refactor xcom setting variables
- added comments

valayDave committed:
Changed retry setting.

valayDave committed:
Fixed bug when using catch + retry

valayDave committed:
making changes sync with master

valayDave committed:
Merge branch 'master' into aft-mm

valayDave committed:
Fixing logs related change from master.

valayDave committed:
Removed todos + added some validation logic.

valayDave committed:
Added support for Parameters.
- Supporting int, str, bool, float, JSONType

valayDave committed:
Removed hacky line.

valayDave committed:
Added hacky line

valayDave committed:
Added support for Branching with Airflow
- remove `next` function in `AirflowTask`
- `AirflowTask`s has no knowledge of next tasks.
- removed todos + added some todos
- Graph construction on airflow side using graph_structure datastructure.
- graph_structure comes from`FlowGraph.output_steps()[1]`

valayDave committed:
fixing bugs in service account setting

valayDave committed:
2 changes :
- removed hacky line
- added support to directly throw dags in s3.

valayDave committed:
Adding exceptions for missing features.

valayDave committed:
Max workers and worker pool support.

valayDave committed:
@project, @schedule, default active dag support.
- Added a flag to allow setting dag as active on creation
- Airflow compatible schedule interval
- Project name fixes.

valayDave committed:
Bug fix in active defaults.

valayDave committed:
bug fix.

valayDave committed:
Bug fix.

valayDave committed:
Added secrets support.

valayDave committed:
Bug fixes WRT Kubernetes secrets + k8s deployments.
- Fixing some error messages.
- Added some comments.

valayDave committed:
Fixing bug fix in schedule.

valayDave committed:
Removing un-used code paths + code cleanup

valayDave committed:
Changes based on PR comments:
- removed airflow xcom push file , moved to decorator code
- removed prefix configuration
- nit fixes.

valayDave committed:
Moved exceptions to airflow compiler

valayDave committed:
Removing un-used args.

valayDave committed:
shortened length of run-id and task-id hashes.

valayDave committed:
changed `--is-active`->`--is-paused-upon-creation`
- dags are active by default.

valayDave committed:
Made file path required arguement.

valayDave committed:
Remove readme.

valayDave committed:
Removed usless code WRT project decorator.

valayDave committed:
Fixing mf log related bug
- No double logging on metaflow.

valayDave committed:
Added validations about:
- un-supported decorators
- foreach
Changed where validations are done to not save the package.

valayDave committed:
Refactored code .
- removed compute/k8s.py
- Moved k8s code to airflow_compiler.py
- ran isort to airflow_compiler.py

valayDave committed:
Fixed passing secrets with kubernetes.

valayDave committed:
3 changes:
- Removing s3 dep
- remove uesless import
- added `deployed_on` in dag file template

valayDave committed:
reverting change.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow-tests

valayDave committed:
ran black,

valayDave committed:
Removing un-used code / resolved-todos.

valayDave committed:
Merge branch 'airflow-tests' into airflow

valayDave committed:
variable renamemetaflow_specific_args

valayDave committed:
Option help bug fix.

valayDave committed:
fixed printing bug

valayDave committed:
Addressed comments.

valayDave committed:
Airflow sensor api (#3)

* Fixed run-id setting
- Change gaurentees that multiple dags triggered at same moment have unique run-id

* added allow multiple in `Decorator` class

* Airflow sensor integration.
 >> support added for :
- ExternalTaskSensor
- S3KeySensor
- SqlSensor
>> sensors allow multiple decorators
>> sensors accept those arguments which are supported by airflow

* Added `@airflow_schedule_interval` decorator
* Fixing bug run-id related in env variable setting.

valayDave committed:
porting savin's comments
- next changes : addressing comments.

valayDave committed:
Removed Sensor metadata extraction.

valayDave committed:
Multiple Changes based on comments:
1. refactored `create_k8s_args` into _to_job
2. Addressed comments for snake casing
3. refactored `attrs` for simplicity.
4. refactored `metaflow_parameters` to `parameters`.
5. Refactored setting of `input_paths`

valayDave committed:
Following Changes:
- Refactors setting scheduling interval
- refactor dag file creating function
- refactored is_active to is_paused_upon_creation
- removed catchup

valayDave committed:
Fixed defaults passing
- Addressed comments for airflow.py

valayDave committed:
nit fixes to dag.py's templating variables.

valayDave committed:
nit fix : variable name change.

valayDave committed:
added mising Run-id prefixes to variables.
- merged `hash` and `dash_connect` filters.

valayDave committed:
Merge branch 'master' into af-mmr

valayDave committed:
squashing bugs after changes from master.

valayDave committed:
Moving information from airflow_utils to compiler (#56)

- commenting todos to organize unfinished changes.
- some environment variables set via`V1EnvVar`
    - `client.V1ObjectFieldSelector` mapped env vars were not working in json form
    - Moving k8s operator import into its own function.
    - env vars moved.

valayDave committed:
ran black.

valayDave committed:
Refactored sensors to different files + bug fix
- bug caused due `util.compress_list`.
- The function doesn't play nice with strings with variety of characters.
- Ensured that exceptions are handled appropriately.
- Made new file for each sensor under `airflow.sensors` module.

valayDave committed:
GPU support added + container naming same as argo

valayDave committed:
Removed unused `sanitize_label` function

valayDave committed:
removed un-used function

valayDave committed:
k8s bug fix

valayDave committed:
Addressing Final comments.  (#57)

- Added dag-run timeout.
- airflow related scheduling checks in decorator.
- Auto naming sensors if no name is provided
- Annotations to k8s operators
- fix: argument serialization for `DAG` arguments (method names refactored like `to_dict` became `serialize`)
- annotation bug fix
- setting`workflow-timeout` for only scheduled dags

savingoyal commented on metaflow/decorators.py:
```diff
@@ -100,6 +100,8 @@ class Decorator(object):
     name = "NONAME"
     defaults = {}
+    # `allow_multiple` allows setting many decorators of the same type to a step/flow.
+    allow_multiple = False
```
We may not need this functionality anymore at the flow level. Can you check with Asad if they still need multiple sensors of the same type for a flow?

valayDave replied:
We need this at the moment so keeping the change.

savingoyal commented on metaflow/plugins/__init__.py:
```diff
@@ -153,7 +157,22 @@ def get_plugin_cli():
+
+FLOW_DECORATORS = [CondaFlowDecorator, ScheduleDecorator, ProjectDecorator] + [
+    S3KeySensorDecorator,
+    SQLSensorDecorator,
```
We can remove `SQLSensor` for now. One less thing to maintain.

valayDave replied:
We also need this so Keeping ti too

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,198 @@
+    obj.dag_name = resolve_dag_name(obj, name)
+
+
+@airflow.command(help="Compile a new version of this workflow to Airflow DAG.")
```
nit - `flow` instead of `workflow`

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,198 @@
+    "--namespace",
+    "user_namespace",
+    default=None,
+    # TODO (savin): Identify the default namespace?
```
Can you remind me what is the namespace in which this flow executes?

valayDave replied:
Currently, it is the `user`'s namespace. The user who is creating the DAG file. So for a CI system, it is the username of the CI system.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,198 @@
+    "--workflow-timeout",
+    default=None,
+    type=int,
+    help="Workflow timeout in seconds. Enforced only for scheduled DAGs.",
```
Can we throw a pretty error to the user if we detect that they want to set the workflow timeout for an unscheduled DAG?

valayDave replied:
Change included in `_validate_workflow`

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,198 @@
+        file,
+    )
+    with open(file, "w") as f:
+        f.write(flow.compile())
```
Can you add a check ensuring that `file` isn't the same as the flow file?

valayDave replied:
Change included in `_validate_workflow`

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,198 @@
+
+
+def resolve_dag_name(obj, name):
+    project = current.get("project_name")
```
Are there any length limitations for dag names?

valayDave replied:
Under Review

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,120 @@
+    def flow_init(
+        self, flow, graph, environment, flow_datastore, metadata, logger, echo, options
+    ):
+        schedule_interval = flow._flow_decorators.get("airflow_schedule_interval")
```
I haven't looked deeply into the `flow_decorators` implementation yet, but would this work all the time? Is the `_flow_decorators` object fully resolved when this hook gets executed?

savingoyal replied:
Maybe better to move this check out of this decorator and into the `_validate_workflow` or within `airflow.py`.

valayDave replied:
Flow has `_flow_decorators` resolved before `flow_init`. Actually to call `flow_init` we do `flow._flow_decorators.values()`. `flow_init` is called before the execution reaches airflow_cli so the validation will be done correctly.

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,120 @@
+                "Flow cannot have @schedule and @airflow_schedule_interval at the same time. Use any one."
+            )
+
+        self._option_values = options
```
What/where is this being used?

valayDave replied:
`self._option_value` holds the value to the `--schedule` CLI option introduced by this flow decorator. Its used to set `self.schedule`.

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,120 @@
+)
+
+
+class AirflowScheduleIntervalDecorator(FlowDecorator):
```
Can you remind me why we need this in addition to `@schedule`? It seems that they are practically the same at this point (except for the returned values which can be aligned)?

valayDave replied:
Adding arguments from CLI. `@schedule` doesn't introduce any options related to scheduling in the cli. This is also like syntactic sugar. It can be removed.

valayDave replied:
TODO : Remvoe schedule interval

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,120 @@
+    ):
+        # find out where the execution is taking place.
+        # Once figured where the execution is happening then we can do
+        # handle xcom push / pull differently
```
Can you expand on this comment? What is the different mode of handling xcoms?

valayDave replied:
Different operators push xcom differently. This comment was added right at the start of the integration when we could potentially integrate into other operators like BashOperator / BatchOperator etc.

valayDave replied:
Removed this comment. Resolving conversation

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,120 @@
+        json.dump(xcom_dict, f)
+
+
+AIRFLOW_STATES = dict(
```
It doesn't seem like these values are being used anywhere here.

valayDave replied:
Used in ExternalTaskSensor

savingoyal commented on metaflow/plugins/airflow/sensors/__init__.py:
```diff
@@ -0,0 +1,10 @@
+SUPPORTED_SENSORS = [
+    ExternalTaskSensorDecorator.name,
+    S3KeySensorDecorator.name,
+    SQLSensorDecorator.name,
```
We can consider sunsetting SQLSensor for now (pending confirmation from Asad)

valayDave replied:
Confirmed that we need the sensor.

savingoyal commented on metaflow/plugins/airflow/sensors/base_sensor.py:
```diff
@@ -0,0 +1,76 @@
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Is the task is task name a metaflow task?
```
remove this comment

savingoyal commented on metaflow/plugins/airflow/sensors/base_sensor.py:
```diff
@@ -0,0 +1,76 @@
+
+    def compile(self):
+        """
+        compile the arguments for `airflow create` command.
```
Compile

savingoyal commented on metaflow/plugins/airflow/sensors/base_sensor.py:
```diff
@@ -0,0 +1,76 @@
+    def compile(self):
+        """
+        compile the arguments for `airflow create` command.
+        This will even check if the arguments are acceptible.
```
There doesn't seem to be any compilation or checks happening in this method?

valayDave replied:
renamed this function

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+
+
+class ExternalTaskSensorDecorator(AirflowSensorDecorator):
+    operator_type = SensorNames.EXTERNAL_TASK_SENSOR
```
I am not sure why SensorNames are defined in a separate utility file. To follow the existing conventions, you can simply have 
```
TYPE = 'external_task_sensor'
```
and reference `ExternalTaskSensorDecorator.TYPE` everywhere.

valayDave replied:
Some Context: each airflow sensor-related decorator creates an `AirflowTask`(internal object) from the decorator. This object is serialized and is part of the configuration in `dag.py`. Each AirflowTask also has an operator_type. These map to `SensorNames`. We reuse these names to map to the right airflow construct. This code is present airflow_utils and can't move it from there.

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+        failed_states=None,
+        execution_delta=None,
+        check_existence=True,
+        # we cannot add `execution_date_fn` as it requires a python callable.
```
We

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+        # we cannot add `execution_date_fn` as it requires a python callable.
+        # Passing around a python callable is non-trivial since we are passing a
+        # callable from metaflow-code to airflow python script. In this conversion we cannot
+        # that the callable will as the user expects since we cannot transfer dependencies
```
It seems that the last sentence is missing a few words.

valayDave replied:
fixed.

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+                    )
+                )
+        else:
+            self.attributes["allowed_states"] = ["success"]
```
Shouldn't this be `AIRFLOW_STATES.success`?

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+        **AirflowSensorDecorator.defaults,
+        external_dag_id=None,
+        external_task_ids=None,
+        allowed_states=["success"],
```
`AIRFLOW_STATES.success`

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+        if type(self.attributes["allowed_states"]) == str:
+            if self.attributes["allowed_states"] not in list(AIRFLOW_STATES.values()):
+                raise AirflowException(
+                    "`%s` is an invalid input for the `%s` argument of `@%s`. Accepted values are : %s"
```
x is an invalid value of y for z. Accepted values are foo, bar, baz.

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+            ]
+            if len(enum_not_matched) > 0:
+                raise AirflowException(
+                    "`%s` is an invalid input for the `%s` argument of `@%s`. Accepted values are : %s"
```
Same comment as above

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+        if self.attributes["execution_delta"] is not None:
+            if not isinstance(self.attributes["execution_delta"], timedelta):
+                raise AirflowException(
+                    "`%s` argument of `@%s` does not accept values of type `%s`. Accepted type of `%s` is `%s`"
```
Same comment as above.

savingoyal commented on metaflow/plugins/airflow/sensors/external_task_sensor.py:
```diff
@@ -0,0 +1,96 @@
+                        self.name,
+                        str(type(self.attributes["execution_delta"])),
+                        "execution_delta",
+                        "datetime.timedelta",
```
It seems that only 2 values need to be parameterized instead of 5.

savingoyal commented on metaflow/plugins/airflow/sensors/s3_sensor.py:
```diff
@@ -0,0 +1,26 @@
+
+class S3KeySensorDecorator(AirflowSensorDecorator):
+    name = "airflow_s3_key_sensor"
+    operator_type = SensorNames.S3_SENSOR
```
Same comment as the previous file

valayDave replied:
https://github.com/valayDave/metaflow/pull/58#discussion_r898587618

savingoyal commented on metaflow/plugins/airflow/sensors/s3_sensor.py:
```diff
@@ -0,0 +1,26 @@
+    def compile(self):
+        if self.attributes["bucket_key"] is None:
+            raise AirflowException(
+                "`%s` argument of `@%s`cannot be `None`." % ("bucket_key", self.name)
```
`bucket_key for foo can't be empty.`

savingoyal commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -337,3 +338,32 @@ def _save_package_once(cls, flow_datastore, package):
             )[0]
+
+    @classmethod
+    def _parse_decorator_spec(cls, deco_spec):
```
Why is this needed?

valayDave replied:
This was to support secrets. I will check if it works after removing this

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+    headline = "Sensor package not found"
+
+
+LABEL_VALUE_REGEX = re.compile(r"^[a-zA-Z0-9]([a-zA-Z0-9\-\_\.]{0,61}[a-zA-Z0-9])?$")
```
Not needed anymore.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+LABEL_VALUE_REGEX = re.compile(r"^[a-zA-Z0-9]([a-zA-Z0-9\-\_\.]{0,61}[a-zA-Z0-9])?$")
+
+TASK_ID_XCOM_KEY = "metaflow_task_id"
+RUN_ID_LEN = 12
```
`RUN_ID_PREFIX` and `SEPARATOR` combine to 8 chars. Should we consider increasing the size of `RUN_ID`

savingoyal replied:
On further review, it seems that `RUN_ID_LEN` is not 12 but 20. You can either choose a better name or bump the `RUN_ID_LEN` to 20 and account for the prefix in the id generator.

valayDave replied:
`RUN_ID_LEN` is the length of the hash and not the full string. I can rename that to `RUN_HASH_LEN`

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+        return list(cls.__dict__.values())
+
+
+def run_id_creator(val):
```
can we combine `task_id_creator` and `run_id_creator`? They seem to be practically the same modulo length.

valayDave replied:
I can remove the functions and make it one function but the airflow macro should ideally be different. `run_id_creator` is used like [this](https://github.com/valayDave/metaflow/blob/148549cc6adf18a1b7adeb0821f2002fdf7aeb39/metaflow/plugins/airflow/airflow.py#L341)

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+)
+
+
+class SensorNames:
```
This may not be needed.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+    # `_arg_types` is a dictionary which represents the types of the arguments of an Airflow `DAG`.
+    # `_arg_types` is used when parsing types back from the configuration json.
+    # It doesn't cover all the arguments but covers many of the important one which can come from the cli.
+    _arg_types = {
```
Do we need to list all args? `catchup`, `tags`, `depends_on_past`, `email*`, `retr*`, `queue`, `priority_*`, `wait_for_downstream`, `sla`, `trigger_rule`, `start_date`?

valayDave replied:
We can remove many. We mainly care about datetime and timedeltas.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+            "priority_weight": int,
+            "wait_for_downstream": bool,
+            "sla": timedelta,
+            "execution_timeout": timedelta,
```
What's the difference between `execution_timeout` and `dagrun_timeout`?

valayDave replied:
Execution timeout is at task level and dagrun_timeout is at dag-run level. Hence `execution_timeout` is in `default_args`. `default_args` contain arguments which are defaults for all tasks in dag.

valayDave replied:
Add comment on where we use defaults which are in parity with batch/stepfn/argo etc.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+        def parse_args(dd):
+            data_dict = {}
+            for k, v in dd.items():
+                # see the comment below for `from_dict`
```
remove the comment

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+
+    @classmethod
+    def deserialize(cls, data_dict):
+        def parse_args(dd, type_check_dict):
```
this can be simplified further? It seems that we have just one element which is `timedelta`.

valayDave replied:
we have more than one `timedelta` / `datetime` elements. Simplified these functions a little.

valayDave replied:
Check if `start_date` is needed.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+        {
+            "secrets": secrets,
+            # Question for (savin):
+            # Default timeout in airflow is 120. I can remove `startup_timeout_seconds` for now. how should we expose it to the user?
```
which timeout is this?

valayDave replied:
`startup_timeout_seconds` – timeout in seconds to startup the pod.

valayDave replied:
Check if this can be set at a server level ?

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+            # Default timeout in airflow is 120. I can remove `startup_timeout_seconds` for now. how should we expose it to the user?
+        }
+    )
+    # Below cannot be passed in dictionary form. After trying a few times it didin't work.
```
What was the error?

valayDave replied:
No error. It was just not recognizing the dictionary. Particularly dictionaries which have the model of `V1ObjectFieldSelector` were not being registered as environment variables.  Because of that, we were not able to pass metadata about `METAFLOW_KUBERNETES_POD_ID` etc. I also tested with a dictionary that had defaults set to the model of `V1EnvVarSource` with `V1ObjectFieldSelector` but it also didn't work. 

BTW it was easily accepting`V1EnvVar1` in JSON dictionary form but anything with `V1EnvVarSource`, wasn't accepted.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+    # to parse dictionaries to `airflow.providers.cncf.kubernetes.backcompat.pod.Resources` object via
+    # `airflow.providers.cncf.kubernetes.backcompat.backward_compat_converts.convert_resources` function
+    resources = args.get("resources")
+    args["resources"] = client.V1ResourceRequirements(
```
GPUs?

valayDave replied:
GPU's are included in the compilation of resources in the `Airflow`(airflow.py) object

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+        return cls(**parse_args(data_dict, cls._arg_types))
+
+
+def _kubernetes_pod_operator_args(operator_args):
```
TODO (savin): Review that this is functionally equivalent to `argo-workflows` and `kubernetes`

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+        # 12 Jan 4:00 PM PDT then "DagB"(current DAG)'s task sensor will try to
+        # look for a "DagA" that got executed at 12 Jan 4:00 PM PDT **exactly**.
+        # They also support a `execution_timeout` argument to
+        from airflow.sensors.external_task_sensor import ExternalTaskSensor
```
Currently, it seems that if I have to define a new sensor, I need to make changes in many different places. Can you look into consolidating all this logic and mappings in a single spot?

valayDave replied:
I can do it more dynamically by importing these sensors via an `importlib` and removing the SensorName stuff altogether. I felt that was overkill so I kept it this way. We can modify this in the near future but for now, we can keep it like this as the integration is not catering to too many sensors.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+        }
+
+    @classmethod
+    def from_dict(cls, jsd, flow_name=None):
```
jsd?

valayDave replied:
fixed variable name

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,410 @@
+                )
+            except ImportError as e:
+                raise KubernetesProviderNotFound(
+                    "This DAG requires a `KubernetesPodOperator`. "
```
`This DAG utilizes KubernetesPodOperator. Please install the Airflow Kubernetes provider using '%s -m pip install apache-airflow-providers-cncf-kubernetes'` % sys.executable

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+
+    parameter_macro = "{{ params | json_dump }}"
+
+    task_id = AIRFLOW_TASK_ID_TEMPLATE_VALUE
```
Why wouldn't you directly use `AIRFLOW_TASK_ID_TEMPLATE_VALUE`. Maybe rename it to `AIRFLOW_TASK_ID`

valayDave replied:
Can be refactored.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+    parameter_macro = "{{ params | json_dump }}"
+
+    task_id = AIRFLOW_TASK_ID_TEMPLATE_VALUE
+    task_id_arg = "--task-id %s" % task_id
```
Is it necessary to define it here given it's only used at one place?

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+        % (RUN_ID_PREFIX, str(RUN_ID_LEN))
+    )
+    # We do echo -n because emits line breaks and we dont want to consider that since it we want same hash value when retrieved in python.
+    run_id_arg = "--run-id %s" % run_id
```
is it necessary to define this variable?

valayDave replied:
The code is organized this way because it was easier for me to read during development. Since we are constructing many things in the cli via macro I was a little inclined to keep everything at one place in variables to understand what is going into the cli.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+    )
+    # We do echo -n because emits line breaks and we dont want to consider that since it we want same hash value when retrieved in python.
+    run_id_arg = "--run-id %s" % run_id
+    attempt = "{{ task_instance.try_number - 1 }}"
```
same comment as above

valayDave replied:
https://github.com/valayDave/metaflow/pull/58#discussion_r898531869

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+        self.worker_pool = worker_pool
+        self.is_paused_upon_creation = is_paused_upon_creation
+        self.workflow_timeout = workflow_timeout
+        self._set_scheduling_interval()
```
`self.schedule = self._set_schedule()`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+
+    def _get_airflow_schedule_interval(self):
+        schedule_interval = self.flow._flow_decorators.get("airflow_schedule_interval")
+        if schedule_interval is None:
```
it seems that we may not need both `schedule_interval` and `schedule`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+        retry_decos = [deco for deco in node.decorators if deco.name == "retry"]
+        if len(retry_decos) > 0:
+            retry_mins = retry_decos[0].attributes["minutes_between_retries"]
+            return timedelta(minutes=int(retry_mins))
```
Do we need to convert to `timedelta`?

valayDave replied:
Yes. This is serialized to a dictionary and reserialzed to timedelta in airflow_utils.py.  

This is the datastructure required by BaseOperator

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+            # Airflow requires defaults set for parameters.
+            if "default" not in param.kwargs:
+                raise MetaflowException(
+                    "The parameter *%s* does not have a "
```
`Parameter foo does...`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+                raise MetaflowException(
+                    "The parameter *%s* does not have a "
+                    "default set. "
+                    "A default is required for parameters when deploying on Airflow."
```
`A default value..`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+            if "default" not in param.kwargs:
+                raise MetaflowException(
+                    "The parameter *%s* does not have a "
+                    "default set. "
```
`default value`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+            parameters.append(dict(name=param.name, value=value))
+
+            # Setting airflow related param args.
+            param_type = param.kwargs.get("type", None)
```
`.get('foo')` is equivalent to `.get('foo', None)`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+            airflow_param = dict(
+                name=param.name,
+            )
+            param_help = param.kwargs.get("help", None)
```
same comment as above

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+                airflow_param["default"] = value
+            if param_help:
+                airflow_param["description"] = param_help
+            if param_type is not None:
```
What is the default param_type is none is specified by the user?

valayDave replied:
Where ever the type of the default value.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+                    airflow_param["type"] = type_transform_dict[JSONTypeClass.name]
+                elif param_type.__name__ in type_transform_dict:
+                    airflow_param["type"] = type_transform_dict[param_type.__name__]
+                    if param_type.__name__ in type_parser and value is not None:
```
Does this line need to be nested within the `elif` block?

valayDave replied:
Refactored here for simplicity.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+                        )
+
+            airflow_params.append(airflow_param)
+        self.parameters = airflow_params
```
let's avoid these hidden assignments.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+            airflow_params.append(airflow_param)
+        self.parameters = airflow_params
+
+        return parameters
```
Why do you need `parameters`? Can't we just reuse `airflow_params`? It will make for a much more maintainable code base.

valayDave replied:
Good catch! I can surely try that for setting parameters at the flow start. 

The purpose of parsing parameters this way is to pass those to airflow. They are important in instantiating `airflow.Param` which in turn populates the parameter macro.

valayDave replied:
Refactored a lot of parameter setting logic. Made it much more simpler.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,670 @@
+
+        return parameters
+
+    def _make_input_path_compressed(
```
TODO(savin): Review rest of the file.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Bug fix from master merge.

valayDave committed:
addressing  savins comments. (#59)

- Added many adhoc changes based for some comments.
- Integrated secrets and `KUBERNETES_SECRETS`
- cleaned up parameter setting
- cleaned up setting of scheduling interval
- renamed `AIRFLOW_TASK_ID_TEMPLATE_VALUE` to `AIRFLOW_TASK_ID`
- renamed `AirflowSensorDecorator.compile` to `AirflowSensorDecorator.validate`
- Checking if dagfile and flow file are same.
- fixing variable names.
- checking out `kubernetes_decorator.py` from master (6441ed5)
- bug fixing secret setting in airflow.
- simplified parameter type parsing logic
- refactoring airflow argument parsing code.

valayDave committed:
refactored an error string

valayDave committed:
refactor :
-`RUN_ID_LEN` to `RUN_HASH_ID_LEN`
- `TASK_ID_LEN` to `TASK_ID_HASH_LEN`

valayDave committed:
Refactored an `id_creator` method.

valayDave committed:
added comment on need for `start_date`

valayDave committed:
Refactored parameter macro settings. (valayDave/metaflow#60)

valayDave committed:
simplifying run-id macro.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow
